### PR TITLE
[Infra] Migrate rest of linux builder workflows off GCP runners.

### PR DESCRIPTION
**Pull request by saienduri**
This commit is part of this larger issue that is tracking our migration off the GCP runners, storage buckets, etc: https://github.com/iree-org/iree/issues/18238.

This builds on https://github.com/iree-org/iree/pull/18381, which migrated
* `linux_x86_64_release_packages`
* `linux_x64_clang_debug`
* `linux_x64_clang_tsan`

Here, we move over the rest of the critical linux builder workflows off of the GCP runners:
* `linux_x64_clang`
* `linux_x64_clang_asan`

This also drops all CI usage of the GCP cache (`http://storage.googleapis.com/iree-sccache/ccache`). Some workflows now use sccache backed by Azure Blob Storage as a replacement. There are few issues with this (https://github.com/mozilla/sccache/issues/2258) that prevent us providing read only access to the cache in PRs created from forks, so **PRs from forks currently don't use the cache and will have slower builds**. We're covering for this slowdown by using larger runners, but if we can roll out caching to all builds then we might use runners with fewer cores.

Along with the changes to the cache, usage of Docker is rebased on images in the https://github.com/iree-org/base-docker-images/ repo and the `build_tools/docker/docker_run.sh` script is now only used by unmigrated workflows (`linux_arm64_clang` and `build_test_all_bazel`).

### CODE REVIEW

File: .github/workflows/ci_linux_x64_clang.yml [hunk 1]
```diff
@@ -28,37 +28,49 @@ jobs:
...
     if: contains(fromJson(needs.setup.outputs.enabled-jobs), 'linux_x64_clang')
-    runs-on:
-      - self-hosted # must come first
-      - runner-group=${{ needs.setup.outputs.runner-group }}
-      - environment=${{ needs.setup.outputs.runner-env }}
-      - cpu
-      - os-family=Linux
+    runs-on: azure-linux-scale
+    container:
+      image: ghcr.io/iree-org/cpubuilder_ubuntu_jammy_x86_64@sha256:2b2ad51d7de988be13086bc618d89d2ba47fbf09eb5b38c60dce82b595fb1c74
```
Comment (ScottTodd) [inline .github/workflows/ci_linux_x64_clang.yml:33]: I'll want to make sure all of these are using the latest image prior to merging. Right now there are a few different versions of the image in use.

(still pinned to a SHA, not the `main` tag)

File: .github/workflows/ci_linux_x64_clang_tsan.yml [hunk 2]
```diff
@@ -32,21 +30,28 @@ jobs:
...
-          ./build_tools/github_actions/docker_run.sh \
-            --env "IREE_CCACHE_GCP_TOKEN=$(gcloud auth application-default print-access-token)" \
-            --env "IREE_WRITE_REMOTE_CCACHE=${IREE_WRITE_REMOTE_CCACHE}" \
-            --env "CCACHE_NAMESPACE=base-bleeding-edge@sha256:14200dacca3a0f3a66f8aa87c6f64729b83a2eeb403b689c24204074ad157418" \
-            gcr.io/iree-oss/base-bleeding-edge@sha256:cf2e78194e64fd0166f4141317366261d7a62432b72e9a324cb8c2ff4e1a515a \
-            ./build_tools/cmake/build_and_test_tsan.sh
+          # Use a modern clang explicitly.
+          CC: clang-19
+          CXX: clang++-19
+        run: ./build_tools/cmake/build_and_test_tsan.sh
```
Comment (ScottTodd) [inline .github/workflows/ci_linux_x64_clang_tsan.yml]: 60 second timeout running the `iree/tools/test/compile_to_continuation.mlir.test` test 🤔 

https://github.com/iree-org/iree/actions/runs/10838630188/job/30077305097?pr=18511#step:5:12673

Can see if that keeps happening or if it was a flake. TSan is still a nightly build so lower priority to fix individual tests like that

Comment (ScottTodd) [reply .github/workflows/ci_linux_x64_clang_tsan.yml]: At least it succeeded on a retry.

Comment (saienduri) [reply .github/workflows/ci_linux_x64_clang_tsan.yml]: Popped up again btw

File: test.txt
Comment (ScottTodd) [inline test.txt:1]: Revert before merge. (This should be tested in a separate PR, test commits should not go directly on the branch we intend to merge)

Comment (ScottTodd) [reply test.txt:1]: (Especially if the test commits don't pass the DCO or pre-commit checks)

File: .github/workflows/ci_linux_x64_clang_tsan.yml (hunk 2 above)
Comment (ScottTodd) [reply .github/workflows/ci_linux_x64_clang_tsan.yml]: Still timing out. Looks like several long tests are quite a bit slower on these runners.

This PR:
https://github.com/iree-org/iree/actions/runs/10852963659/job/30120192675?pr=18511#step:5:12518
```
1455/1469 Test  #987: iree/tests/compiler_driver/streams.mlir.test ......................................................................   Passed   10.64 sec
1456/1469 Test    #5: iree/tools/test/compile_to_phase.mlir.test ........................................................................   Passed   20.14 sec
1457/1469 Test #1227: iree/tests/e2e/stablehlo_models/edge_detection.mlir.test ..........................................................   Passed   11.17 sec
1458/1469 Test #2127: iree/samples/multiple_modules/pipeline_sync.mlir.test .............................................................   Passed    8.78 sec
1459/1469 Test #1226: iree/tests/e2e/stablehlo_models/collatz.mlir.test .................................................................   Passed   11.26 sec
1460/1469 Test  #323: iree/compiler/Codegen/LLVMGPU/test/ROCDL/pipeline_vector_distribute.mlir.test .....................................   Passed   17.47 sec
1461/1469 Test #2126: iree/samples/multiple_modules/pipeline_async.mlir.test ............................................................   Passed    9.55 sec
1462/1469 Test #2115: iree/samples/custom_dispatch/cpu/mlp_plugin/mlp_linalg_two_matmul.mlir.test .......................................   Passed   10.58 sec
1463/1469 Test  #287: iree/compiler/Codegen/LLVMGPU/test/nvvm_pipeline_test.mlir.test ...................................................   Passed   28.12 sec
1464/1469 Test #1230: iree/tests/e2e/stablehlo_models/mnist_fake_weights.mlir.test ......................................................   Passed   22.33 sec
1465/1469 Test #1229: iree/tests/e2e/stablehlo_models/fullyconnected.mlir.test ..........................................................   Passed   25.03 sec
1466/1469 Test #1150: iree/tests/e2e/regression/libm_linking.mlir.test ..................................................................   Passed   25.27 sec
1467/1469 Test #1231: iree/tests/e2e/stablehlo_models/unidirectional_lstm.mlir.test .....................................................   Passed   31.80 sec
1468/1469 Test  #391: iree/compiler/ConstEval/test/jit_globals.mlir.test ................................................................   Passed   44.56 sec
1469/1469 Test    #4: iree/tools/test/compile_to_continuation.mlir.test .................................................................***Timeout  60.01 sec
```

Baseline:
https://github.com/iree-org/iree/actions/runs/10846625133/job/30099971895#step:3:12573
```
1457/1468 Test  #361: iree/compiler/Codegen/SPIRV/test/pipeline_matmul_promotion.mlir.test ..............................................   Passed   10.68 sec
1458/1468 Test #2125: iree/samples/multiple_modules/pipeline_async.mlir.test ............................................................   Passed    3.73 sec
1459/1468 Test #2126: iree/samples/multiple_modules/pipeline_sync.mlir.test .............................................................   Passed    3.77 sec
1460/1468 Test #2114: iree/samples/custom_dispatch/cpu/mlp_plugin/mlp_linalg_two_matmul.mlir.test .......................................   Passed    4.55 sec
1461/1468 Test  #323: iree/compiler/Codegen/LLVMGPU/test/ROCDL/pipeline_vector_distribute.mlir.test .....................................   Passed   12.30 sec
1462/1468 Test #1229: iree/tests/e2e/stablehlo_models/mnist_fake_weights.mlir.test ......................................................   Passed    9.25 sec
1463/1468 Test #1228: iree/tests/e2e/stablehlo_models/fullyconnected.mlir.test ..........................................................   Passed    9.40 sec
1464/1468 Test #1149: iree/tests/e2e/regression/libm_linking.mlir.test ..................................................................   Passed   10.03 sec
1465/1468 Test #1230: iree/tests/e2e/stablehlo_models/unidirectional_lstm.mlir.test .....................................................   Passed   10.73 sec
1466/1468 Test  #287: iree/compiler/Codegen/LLVMGPU/test/nvvm_pipeline_test.mlir.test ...................................................   Passed   16.55 sec
1467/1468 Test  #391: iree/compiler/ConstEval/test/jit_globals.mlir.test ................................................................   Passed   17.91 sec
1468/1468 Test    #4: iree/tools/test/compile_to_continuation.mlir.test .................................................................   Passed   27.82 sec
```

Comment (ScottTodd) [reply .github/workflows/ci_linux_x64_clang_tsan.yml]: We could skip those specific tests on TSan or increase the ctest timeout. Looks like we currently set it per test, with no easy way to override or scale across the whole ctest run.

**Review by ScottTodd — APPROVED** [review]

File: .github/workflows/ci_linux_x64_clang_tsan.yml (hunk 2 above)
Comment (ScottTodd) [reply .github/workflows/ci_linux_x64_clang_tsan.yml]: Increased the timeouts.

Comment (saienduri) [reply .github/workflows/ci_linux_x64_clang_tsan.yml]: Thanks Scott!
